### ISSUE TITLE
RSS ArgumentsOutOfRangeException

### ISSUE THREAD

**Issue by niebl** [niebl/BTDS-project#9]
circumstances are identical to issue #8

How to replicate:
1. Let a person enter the FOV of the camera until they are registered on screen
2. Let an additional person enter the FOV to be registered
3. Let the first person leave the FOV while the second person remains

Desired Action: Sensor keeps sending the coordinates of Person 2 to the server, while Person 1 is "forgotten".
Observed Action: Sensor is stuck in sending the coordinates of Person 1 to the server while disregarding the coordinates of Person 2.

**Comment by niebl** [comment]
after some experimenting it turned out that upon entering of the second body, the coordinates of the inhabitants in the Array `Tracked_Inhabitants[]` are seemingly "swapped".
Meaning that upon entering of the second Person, the second Tracked_Inhabitant object is marked as true, but the coordinates of the second person are written into the first object and the coordinates of the first person are written into the second object.

**Comment by niebl** [comment]
The reason may be the fact that tracked_bodies, a `List<Body>` from the kinect is used to fill the tracked_inhabitant objects. It could be that new bodies are not appended to the list, but prepended, causing the "swap"

```C#
        private void bodies_To_Inhabitants(List<Body> tracked_bodies)
        {
            //the loop assigning bodies_ids is not necessary if DrawTracked_Bodies is called before

            for(int inhabIndex = 0; inhabIndex < 6; inhabIndex++)
            {
                tracked_Inhabitants[inhabIndex].tracked = (bodies_ids[inhabIndex] != 0);

                if(tracked_Inhabitants[inhabIndex].tracked) //if the body is currently tracked by the sensor
                {
                    var current_body = tracked_bodies[inhabIndex].Joints[JointType.SpineMid];

                    //fill the Objects with the correct coordinates
                    tracked_Inhabitants[inhabIndex].x = Math.Round(current_body.Position.Z, 2);
                    tracked_Inhabitants[inhabIndex].y = Math.Round(current_body.Position.X, 2) * (-1);
                }
                else
                {
                    break;
                }

            }
            
        }
```

**Comment by niebl** [comment]
![image](https://user-images.githubusercontent.com/53115461/129082724-b85d3454-cfc8-4684-bbdb-745b5537645b.png)
![image](https://user-images.githubusercontent.com/53115461/129083030-e650a7c2-fd1f-4c42-9363-eebfbc7d4357.png)

I retract the swap-hypothesis. Testing it again did not bring evidence that a swap is happening.

**Comment by niebl** [comment]
> 
> 
> The reason may be the fact that tracked_bodies, a `List<Body>` from the kinect is used to fill the tracked_inhabitant objects. It could be that new bodies are not appended to the list, but prepended, causing the "swap"
> 
> ```cs
>         private void bodies_To_Inhabitants(List<Body> tracked_bodies)
>         {
>             //the loop assigning bodies_ids is not necessary if DrawTracked_Bodies is called before
> 
>             for(int inhabIndex = 0; inhabIndex < 6; inhabIndex++)
>             {
>                 tracked_Inhabitants[inhabIndex].tracked = (bodies_ids[inhabIndex] != 0);
> 
>                 if(tracked_Inhabitants[inhabIndex].tracked) //if the body is currently tracked by the sensor
>                 {
>                     var current_body = tracked_bodies[inhabIndex].Joints[JointType.SpineMid];
> 
>                     //fill the Objects with the correct coordinates
>                     tracked_Inhabitants[inhabIndex].x = Math.Round(current_body.Position.Z, 2);
>                     tracked_Inhabitants[inhabIndex].y = Math.Round(current_body.Position.X, 2) * (-1);
>                 }
>                 else
>                 {
>                     break;
>                 }
> 
>             }
>             
>         }
> ```

I found the issue! it is caused by the break statement at the end of the loop!
When you take it out a new error emerges.

**Comment by niebl** [comment]
![image](https://user-images.githubusercontent.com/53115461/129086386-b7276538-d394-4651-bdac-50a0f6bb34bd.png)
Understandably, the error is thrown because a corresponding body at tracked_bodies does not exist. A break statement previously was an easy fix, but it does not account for empty fields.

**Comment by niebl** [comment]
We are now back at the same error discussed in issue #8

**Comment by niebl** [comment]
I've been thinking over night.
It makes sense. since tracked_bodies is a list, taking away the first member will take the second member into first place.
I think a quick and dirty approach might be to remember the body ID and do a search over the members of the list. Since the amount of bodies in a scene is 6 at maximum, the runtime shouldn't be of concern

**Comment by niebl** [comment]
This code is working.
Instead of looking for the Body object in tracked_bodies by index, it is taking the corresponding id that was stored in bodies_ids and does a search over the list until it finds the corresponding body.
```C#
        private void bodies_To_Inhabitants(List<Body> tracked_bodies)
        {
            //the loop assigning bodies_ids is not necessary if DrawTracked_Bodies is called before

            for(int inhabIndex = 0; inhabIndex < 6; inhabIndex++)
            {
                tracked_Inhabitants[inhabIndex].tracked = (bodies_ids[inhabIndex] != 0);

                if(tracked_Inhabitants[inhabIndex].tracked) //if the body is currently tracked by the sensor
                {
                    //we search for the body with the corresponding ID
                    Body tracked_body = tracked_bodies.Find(x => x.TrackingId == bodies_ids[inhabIndex]);

                    var current_body = tracked_body.Joints[JointType.SpineMid];

                    //fill the Objects with the correct coordinates
                    tracked_Inhabitants[inhabIndex].x = Math.Round(current_body.Position.Z, 2);
                    tracked_Inhabitants[inhabIndex].y = Math.Round(current_body.Position.X, 2) * (-1);

                }
                else
                {
                    continue;
                }

            }
            
        }
```
Might need some catch block in case it can't find the Body in the future, but I am not sure if that Problem can even appear.

EDIT: elaboration

**Comment by niebl** [comment]
Issue solved in commit 5fa7f76e9d3708e0b9e898824631695eea0275d3
